File: src/skills/testrunner/executor.ts
/**
 * Test executor for running compliance tests.
 * 
 * Orchestrates test execution using the TestRunnerSkill.
 */

import { TestParser } from './parser.js';
import { StrictValidator } from './validator.js';
import type {
  ExecutionOptions,
  TestSpec,
  TestCase,
  TestResult,
  TestSuiteResult,
  HttpResponse,
  AssertionResult,
} from './types.js';

/**
 * Default execution options
 */
const DEFAULT_OPTIONS: ExecutionOptions = {
  baseUrl: 'http://localhost:8765',
  cacheMode: 'read_write',
  temperature: 0.0,
  timeout: 60,
  strictOnly: false,
  parallel: 1,
};

/**
 * Executes compliance tests against an SDK server.
 * 
 * Can run in two modes:
 * 1. Agentic: Uses AI to validate natural language assertions
 * 2. Strict-only: Only runs deterministic assertions (faster, no LLM)
 */
export class TestExecutor {
  private options: ExecutionOptions;
  private results: TestSuiteResult[] = [];

  constructor(options: Partial<ExecutionOptions> = {}) {
    this.options = { ...DEFAULT_OPTIONS, ...options };
  }

  /**
   * Run all tests from a markdown content string.
   */
  async runContent(content: string, filename: string = 'test.md'): Promise<TestSuiteResult> {
    const parser = new TestParser();
    const spec = parser.parse(content, filename);

    return this.runSpec(spec, filename);
  }

  /**
   * Run all tests from a parsed spec.
   */
  async runSpec(spec: TestSpec, filename: string = ''): Promise<TestSuiteResult> {
    // Check tags
    if (this.options.tags && this.options.tags.length > 0) {
      const hasMatchingTag = this.options.tags.some(tag => spec.tags.includes(tag));
      if (!hasMatchingTag) {
        return {
          name: spec.name,
          testFile: filename,
          cases: [],
          passed: 0,
          failed: 0,
          skipped: 1,
        };
      }
    }

    if (this.options.skipTags && this.options.skipTags.length > 0) {
      const hasSkipTag = this.options.skipTags.some(tag => spec.tags.includes(tag));
      if (hasSkipTag) {
        return {
          name: spec.name,
          testFile: filename,
          cases: [],
          passed: 0,
          failed: 0,
          skipped: 1,
        };
      }
    }

    const results: TestResult[] = [];

    for (const testCase of spec.cases) {
      const startTime = Date.now();
      const caseResult = await this.runCase(spec, testCase);
      caseResult.durationMs = Date.now() - startTime;
      results.push(caseResult);
    }

    const passed = results.filter(r => r.passed).length;
    const failed = results.filter(r => !r.passed && r.details !== 'skipped').length;
    const skipped = results.filter(r => r.details === 'skipped').length;

    const suiteResult: TestSuiteResult = {
      name: spec.name,
      testFile: filename,
      cases: results,
      passed,
      failed,
      skipped,
    };

    this.results.push(suiteResult);
    return suiteResult;
  }

  /**
   * Run a single test case.
   */
  private async runCase(_spec: TestSpec, testCase: TestCase): Promise<TestResult> {
    const caseName = testCase.name;

    if (!testCase.request) {
      return {
        name: caseName,
        passed: false,
        assertions: [],
        details: 'skipped',
        durationMs: 0,
      };
    }

    // Make request
    const response = await this.httpRequest(
      testCase.request.method,
      testCase.request.path,
      testCase.request.body,
      testCase.request.headers,
      testCase.request.stream
    );

    // Handle request error
    if (response.error) {
      if (testCase.expected === 'failure') {
        return {
          name: caseName,
          passed: true,
          assertions: [],
          details: 'Expected failure occurred',
          durationMs: 0,
        };
      }
      return {
        name: caseName,
        passed: false,
        assertions: [],
        details: `Request error: ${response.error}`,
        durationMs: 0,
      };
    }

    const assertionResults: AssertionResult[] = [];
    let allPassed = true;

    // Run strict assertions
    if (testCase.strict) {
      const validator = new StrictValidator();
      const strictResult = validator.validate(response, testCase.strict);

      assertionResults.push({
        type: 'strict',
        passed: strictResult.passed,
        details: strictResult.results,
      });

      if (!strictResult.passed) {
        allPassed = false;
      }
    }

    // Run natural language assertions (if not strict-only mode)
    if (!this.options.strictOnly) {
      for (const assertion of testCase.assertions) {
        // In agentic mode, we'd use LLM to validate
        // For now, we just record the assertion for the agent
        assertionResults.push({
          type: 'natural',
          assertion,
          passed: true, // Assume passed unless agent says otherwise
          needsValidation: true,
        });
      }
    }

    // Check expected outcome
    if (testCase.expected === 'failure') {
      // For expected failures, check if we got an error response
      if (response.status >= 400) {
        allPassed = true;
      } else {
        allPassed = false;
      }
    }

    return {
      name: caseName,
      passed: allPassed,
      assertions: assertionResults,
      details: allPassed ? 'All assertions passed' : 'Some assertions failed',
      durationMs: 0,
    };
  }

  /**
   * Make an HTTP request to the SDK server.
   */
  async httpRequest(
    method: string,
    path: string,
    body?: Record<string, unknown>,
    headers?: Record<string, string>,
    stream?: boolean
  ): Promise<HttpResponse> {
    const url = `${this.options.baseUrl}${path}`;

    const requestHeaders: Record<string, string> = {
      'Content-Type': 'application/json',
      ...headers,
    };

    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.options.timeout * 1000);

      const response = await fetch(url, {
        method: method.toUpperCase(),
        headers: requestHeaders,
        body: body ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      const contentType = response.headers.get('content-type') || '';

      if (stream && contentType.includes('text/event-stream')) {
        // Collect SSE chunks
        const chunks: unknown[] = [];
        const reader = response.body?.getReader();
        const decoder = new TextDecoder();

        if (reader) {
          let done = false;
          while (!done) {
            const result = await reader.read();
            done = result.done;

            if (result.value) {
              const text = decoder.decode(result.value, { stream: true });
              const lines = text.split('\n');

              for (const line of lines) {
                const trimmed = line.trim();
                if (trimmed.startsWith('data: ')) {
                  const data = trimmed.slice(6);
                  if (data !== '[DONE]') {
                    try {
                      chunks.push(JSON.parse(data));
                    } catch {
                      chunks.push({ raw: data });
                    }
                  }
                }
              }
            }
          }
        }

        return {
          status: response.status,
          headers: Object.fromEntries(response.headers.entries()),
          body: chunks,
          format: 'sse',
        };
      } else {
        let responseBody: unknown;
        try {
          responseBody = await response.json();
        } catch {
          responseBody = await response.text();
        }

        return {
          status: response.status,
          headers: Object.fromEntries(response.headers.entries()),
          body: responseBody,
        };
      }
    } catch (error) {
      return {
        status: 0,
        headers: {},
        body: null,
        error: error instanceof Error ? error.message : String(error),
        errorType: error instanceof Error ? error.constructor.name : 'Error',
      };
    }
  }

  /**
   * Get all results.
   */
  getResults(): TestSuiteResult[] {
    return this.results;
  }

  /**
   * Get aggregated summary.
   */
  getSummary(): {
    total: number;
    passed: number;
    failed: number;
    skipped: number;
    suites: TestSuiteResult[];
  } {
    const totalPassed = this.results.reduce((sum, s) => sum + s.passed, 0);
    const totalFailed = this.results.reduce((sum, s) => sum + s.failed, 0);
    const totalSkipped = this.results.reduce((sum, s) => sum + s.skipped, 0);

    return {
      total: totalPassed + totalFailed + totalSkipped,
      passed: totalPassed,
      failed: totalFailed,
      skipped: totalSkipped,
      suites: this.results,
    };
  }

  /**
   * Print results to console.
   */
  printResults(): void {
    const summary = this.getSummary();

    console.log('\nCompliance Test Results');
    console.log('='.repeat(50));

    for (const suite of summary.suites) {
      if (suite.skipped > 0 && suite.cases.length === 0) {
        console.log(`\n⊘ ${suite.name}: SKIPPED`);
        continue;
      }

      const status = suite.failed === 0 ? '✓' : '✗';
      console.log(`\n${status} ${suite.name}: ${suite.passed}/${suite.passed + suite.failed} passed`);

      for (const testCase of suite.cases) {
        const caseStatus = testCase.passed ? '✓' : '✗';
        console.log(`  ${caseStatus} ${testCase.name}`);

        if (!testCase.passed && testCase.assertions.length > 0) {
          for (const assertion of testCase.assertions) {
            if (!assertion.passed) {
              const detail = assertion.reason || assertion.assertion || 'Failed';
              console.log(`    → ${detail}`);
            }
          }
        }
      }
    }

    console.log('\n' + '='.repeat(50));
    console.log(`Summary: ${summary.passed}/${summary.passed + summary.failed} tests passed`);
    if (summary.skipped > 0) {
      console.log(`         ${summary.skipped} skipped`);
    }
  }
}

File: compliance/runner/executor.py
"""
Test executor for running compliance tests.

Orchestrates test execution using the TestRunnerSkill.
"""

import asyncio
import time
from pathlib import Path
from typing import Dict, Any, List, Optional
from dataclasses import dataclass


@dataclass
class ExecutionOptions:
    """Options for test execution."""
    base_url: str = "http://localhost:8765"
    cache_mode: str = "read_write"  # read_write, write_only, disabled
    temperature: float = 0.0
    timeout: int = 60
    strict_only: bool = False
    tags: Optional[List[str]] = None
    skip_tags: Optional[List[str]] = None
    parallel: int = 1


class TestExecutor:
    """
    Executes compliance tests against an SDK server.
    
    Can run in two modes:
    1. Agentic: Uses AI to validate natural language assertions
    2. Strict-only: Only runs deterministic assertions (faster, no LLM)
    """
    
    def __init__(self, options: ExecutionOptions):
        self.options = options
        self.results: List[Dict[str, Any]] = []
    
    async def run_file(self, path: Path) -> Dict[str, Any]:
        """
        Run all tests in a single file.
        
        Args:
            path: Path to test file
            
        Returns:
            Test results for the file
        """
        from .parser import TestParser
        from .skill import TestRunnerSkill
        from .validator import StrictValidator
        
        parser = TestParser()
        spec = parser.parse(path)
        
        # Check tags
        if self.options.tags:
            if not any(tag in spec.get("tags", []) for tag in self.options.tags):
                return {
                    "name": spec["name"],
                    "skipped": True,
                    "reason": "No matching tags",
                }
        
        if self.options.skip_tags:
            if any(tag in spec.get("tags", []) for tag in self.options.skip_tags):
                return {
                    "name": spec["name"],
                    "skipped": True,
                    "reason": "Skipped tag",
                }
        
        skill = TestRunnerSkill(
            base_url=self.options.base_url,
            timeout=self.options.timeout,
        )
        
        try:
            results = []
            
            for case in spec.get("cases", []):
                start_time = time.time()
                
                case_result = await self._run_case(skill, spec, case)
                
                case_result["duration_ms"] = (time.time() - start_time) * 1000
                results.append(case_result)
            
            passed = sum(1 for r in results if r.get("passed", False))
            failed = sum(1 for r in results if not r.get("passed", False) and not r.get("skipped", False))
            skipped = sum(1 for r in results if r.get("skipped", False))
            
            return {
                "name": spec["name"],
                "file": str(path),
                "cases": results,
                "passed": passed,
                "failed": failed,
                "skipped": skipped,
            }
        finally:
            await skill.shutdown()
    
    async def _run_case(
        self,
        skill: "TestRunnerSkill",
        spec: Dict[str, Any],
        case: Dict[str, Any],
    ) -> Dict[str, Any]:
        """Run a single test case."""
        from .validator import StrictValidator
        
        case_name = case.get("name", "unnamed")
        request = case.get("request")
        
        if not request:
            return {
                "name": case_name,
                "skipped": True,
                "reason": "No request defined",
            }
        
        # Make request
        response = await skill.http_request(
            method=request.get("method", "POST"),
            path=request.get("path", "/chat/completions"),
            body=request.get("body"),
            headers=request.get("headers"),
            stream=request.get("stream", False),
        )
        
        # Handle request error
        if response.get("error"):
            if case.get("expected") == "failure":
                return {
                    "name": case_name,
                    "passed": True,
                    "reason": "Expected failure occurred",
                }
            return {
                "name": case_name,
                "passed": False,
                "reason": f"Request error: {response.get('error')}",
            }
        
        assertion_results = []
        all_passed = True
        
        # Run strict assertions
        strict = case.get("strict")
        if strict:
            validator = StrictValidator()
            strict_result = validator.validate(response, strict)
            
            assertion_results.append({
                "type": "strict",
                "passed": strict_result["passed"],
                "details": strict_result["results"],
            })
            
            if not strict_result["passed"]:
                all_passed = False
        
        # Run natural language assertions (if not strict-only mode)
        if not self.options.strict_only:
            for assertion in case.get("assertions", []):
                # In agentic mode, we'd use LLM to validate
                # For now, we just record the assertion for the agent
                assertion_results.append({
                    "type": "natural",
                    "assertion": assertion,
                    "passed": None,  # To be validated by agent
                    "needs_validation": True,
                })
        
        # Check expected outcome
        expected = case.get("expected", "success")
        if expected == "failure":
            # For expected failures, check if we got an error response
            if response.get("status", 200) >= 400:
                all_passed = True
            else:
                all_passed = False
        
        return {
            "name": case_name,
            "passed": all_passed,
            "response": {
                "status": response.get("status"),
                "body_preview": str(response.get("body", ""))[:200],
            },
            "assertions": assertion_results,
        }
    
    async def run_directory(self, path: Path) -> Dict[str, Any]:
        """
        Run all tests in a directory.
        
        Args:
            path: Path to test directory
            
        Returns:
            Aggregated test results
        """
        test_files = list(path.glob("**/*.md"))
        
        results = []
        for test_file in test_files:
            result = await self.run_file(test_file)
            results.append(result)
        
        total_passed = sum(r.get("passed", 0) for r in results if not r.get("skipped"))
        total_failed = sum(r.get("failed", 0) for r in results if not r.get("skipped"))
        total_skipped = sum(1 for r in results if r.get("skipped"))
        
        return {
            "total": len(results),
            "passed": total_passed,
            "failed": total_failed,
            "skipped": total_skipped,
            "suites": results,
        }
    
    def print_results(self, results: Dict[str, Any]) -> None:
        """Print results to console."""
        print("\nCompliance Test Results")
        print("=" * 50)
        
        for suite in results.get("suites", []):
            if suite.get("skipped"):
                print(f"\n⊘ {suite['name']}: SKIPPED ({suite.get('reason', '')})")
                continue
            
            status = "✓" if suite.get("failed", 0) == 0 else "✗"
            print(f"\n{status} {suite['name']}: {suite.get('passed', 0)}/{suite.get('passed', 0) + suite.get('failed', 0)} passed")
            
            for case in suite.get("cases", []):
                case_status = "✓" if case.get("passed") else "✗"
                print(f"  {case_status} {case.get('name', 'unnamed')}")
                
                if not case.get("passed") and case.get("assertions"):
                    for assertion in case["assertions"]:
                        if not assertion.get("passed", True):
                            print(f"    → {assertion.get('details', assertion.get('assertion', ''))}")
        
        print("\n" + "=" * 50)
        print(f"Summary: {results.get('passed', 0)}/{results.get('passed', 0) + results.get('failed', 0)} tests passed")
        if results.get("skipped"):
            print(f"         {results.get('skipped')} skipped")
